DATA LOADING

In [1]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [2]:
# Reading in the data
import pandas as pd

df = pd.read_json('jeopardy.json')
df.dtypes

C:\Users\kbell\AppData\Local\Temp\ipykernel_24684\1900014419.py:2: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


category       object
air_date       object
question       object
value          object
answer         object
round          object
show_number     int64
dtype: object

DATA EXPLORATION

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 216930 entries, 0 to 216929
Data columns (total 7 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   category     216930 non-null  object
 1   air_date     216930 non-null  object
 2   question     216930 non-null  object
 3   value        213296 non-null  object
 4   answer       216930 non-null  object
 5   round        216930 non-null  object
 6   show_number  216930 non-null  int64 
dtypes: int64(1), object(6)
memory usage: 13.2+ MB


1. All dtypes but show_number are objects--may need to convert some to integer
2. 216930 rows of data
3. Will focus on 'value' column

DATA PROCESSING

In [4]:
df = df.sort_values('category')
df

category    air_date  \
43802   A JIM CARREY FILM FESTIVAL  2004-12-01   
43797   A JIM CARREY FILM FESTIVAL  2004-12-01   
43807   A JIM CARREY FILM FESTIVAL  2004-12-01   
43813   A JIM CARREY FILM FESTIVAL  2004-12-01   
43819   A JIM CARREY FILM FESTIVAL  2004-12-01   
...                            ...         ...   
79235                “WH”AT IS IT?  2001-06-22   
79224                “WH”AT IS IT?  2001-06-22   
79230                “WH”AT IS IT?  2001-06-22   
79218                “WH”AT IS IT?  2001-06-22   
79213                “WH”AT IS IT?  2001-06-22   

                                                question  value  \
43802  '"Riddle me this, riddle me that, who's afraid...   $400   
43797  '"Blast this Christmas music! It's joyful and ...   $200   
43807  '"Good morning! And in case I don't see you, g...   $600   
43813     '"Hi.  Is there a problem with your service?"'   $800   
43819                   '"Here I come to save the day!"'  $1000   
...                                                  ...    ...   
79235  'You'll find this capital of the Yukon Territo...   $600   
79224  'It's a capricious idea carried out on the spu...   $600   
79230  'Look!  Up in the sky!  It's a helicopter!  No...   $800   
79218  'In a Manhattan cocktail, it's combined with v...   $400   
79213  'In a song from “Snow White”, it's what you do...   $200   

                               answer             round  show_number  
43802                  Batman Forever         Jeopardy!         4658  
43797  How the Grinch Stole Christmas         Jeopardy!         4658  
43807                 The Truman Show         Jeopardy!         4658  
43813                   The Cable Guy         Jeopardy!         4658  
43819                 Man on the Moon         Jeopardy!         4658  
...                               ...               ...          ...  
79235                      Whitehorse  Double Jeopardy!         3885  
79224                            whim  Double Jeopardy!         3885  
79230                      whirlybird  Double Jeopardy!         3885  
79218                         whiskey  Double Jeopardy!         3885  
79213                         whistle  Double Jeopardy!         3885  

[216930 rows x 7 columns]

In [5]:
df = df.reset_index(drop=True)
df

category    air_date  \
0        A JIM CARREY FILM FESTIVAL  2004-12-01   
1        A JIM CARREY FILM FESTIVAL  2004-12-01   
2        A JIM CARREY FILM FESTIVAL  2004-12-01   
3        A JIM CARREY FILM FESTIVAL  2004-12-01   
4        A JIM CARREY FILM FESTIVAL  2004-12-01   
...                             ...         ...   
216925                “WH”AT IS IT?  2001-06-22   
216926                “WH”AT IS IT?  2001-06-22   
216927                “WH”AT IS IT?  2001-06-22   
216928                “WH”AT IS IT?  2001-06-22   
216929                “WH”AT IS IT?  2001-06-22   

                                                 question  value  \
0       '"Riddle me this, riddle me that, who's afraid...   $400   
1       '"Blast this Christmas music! It's joyful and ...   $200   
2       '"Good morning! And in case I don't see you, g...   $600   
3          '"Hi.  Is there a problem with your service?"'   $800   
4                        '"Here I come to save the day!"'  $1000   
...                                                   ...    ...   
216925  'You'll find this capital of the Yukon Territo...   $600   
216926  'It's a capricious idea carried out on the spu...   $600   
216927  'Look!  Up in the sky!  It's a helicopter!  No...   $800   
216928  'In a Manhattan cocktail, it's combined with v...   $400   
216929  'In a song from “Snow White”, it's what you do...   $200   

                                answer             round  show_number  
0                       Batman Forever         Jeopardy!         4658  
1       How the Grinch Stole Christmas         Jeopardy!         4658  
2                      The Truman Show         Jeopardy!         4658  
3                        The Cable Guy         Jeopardy!         4658  
4                      Man on the Moon         Jeopardy!         4658  
...                                ...               ...          ...  
216925                      Whitehorse  Double Jeopardy!         3885  
216926                            whim  Double Jeopardy!         3885  
216927                      whirlybird  Double Jeopardy!         3885  
216928                         whiskey  Double Jeopardy!         3885  
216929                         whistle  Double Jeopardy!         3885  

[216930 rows x 7 columns]

In [6]:
# Using label encoder to convert value dytpe to integer
from sklearn.preprocessing import LabelEncoder

le=LabelEncoder()
df['category'] = le.fit_transform(df['category'])
df['air_date'] = le.fit_transform(df['air_date'])
df['question'] = le.fit_transform(df['question'])
df['value'] = le.fit_transform(df['value'])
df['answer'] = le.fit_transform(df['answer'])
df['round'] = le.fit_transform(df['round'])

In [7]:
# Defining input (X) and predictor (y) variables
from sklearn.model_selection import train_test_split

X = df.drop('value',axis=1)
y = df['value']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25,random_state = True)

MODEL BUILDING & TRAINING - Gaussian

In [8]:
from sklearn.naive_bayes import GaussianNB

model = GaussianNB()

model.fit(X_train,y_train);

MODEL EVALUATION - GAUSSIAN

In [9]:
from sklearn.metrics import (
    accuracy_score,
    f1_score,
    classification_report
)

In [10]:
y_pred = model.predict(X_test)

accuray = accuracy_score(y_pred,y_test)
f1 = f1_score(y_pred,y_test,average='weighted')

print("Accuracy:", accuray)
print("F1 score:", f1)

Accuracy: 0.15459222244758727
F1 score: 0.18434875848994275


In [11]:
#  Trying to get more information on the different indicators of poor performance -- many indicators have division by zero?
print(f'Classication Report: \n{classification_report(y_test,y_pred)}')

Classication Report: 
              precision    recall  f1-score   support

           0       0.00      0.00      0.00       510
           2       0.00      0.00      0.00        15
           4       0.00      0.00      0.00         1
           5       0.00      0.00      0.00       106
           7       0.00      0.00      0.00         1
           9       0.00      0.00      0.00        19
          10       0.00      0.00      0.00         1
          11       0.00      0.00      0.00        57
          12       0.00      0.00      0.00         1
          14       0.00      0.00      0.00       142
          16       0.00      0.00      0.00         1
          17       0.00      0.00      0.00        50
          18       0.00      0.00      0.00        11
          20       0.00      0.00      0.00        47
          22       0.00      0.00      0.00         1
          24       0.00      0.00      0.00         4
          25       0.00      0.00      0.00         6
     

C:\Users\kbell\AppData\Local\anaconda3\envs\Module5\Lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\kbell\AppData\Local\anaconda3\envs\Module5\Lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\kbell\AppData\Local\anaconda3\envs\Module5\Lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f